detection/opencv_detector.go

```diff
@@ -0,0 +1,400 @@
+func (d *OpenCVDetector) initialize() {
+	// Try to load YOLO model files in order of preference
+	modelPaths := []struct {
+		weights string
+		config  string
+		name    string
+	}{
+		{"/opt/yolo/yolov3-tiny.weights", "/opt/yolo/yolov3-tiny.cfg", "YOLOv3-tiny"},
+		{"/opt/yolo/yolov3.weights", "/opt/yolo/yolov3.cfg", "YOLOv3"},
+		{"/opt/yolo/yolov4-tiny.weights", "/opt/yolo/yolov4-tiny.cfg", "YOLOv4-tiny"},
+	}
+
+	classesPath := "/opt/yolo/coco.names"
+
+	// Load class names from COCO dataset
+	// COCO class names reference: https://github.com/pjreddie/darknet/blob/master/data/coco.names
+	// If successful, this loads 80 standard COCO object class names (person, car, dog, etc.)
+	if err := d.loadClassNames(classesPath); err != nil {
+		log.Printf("Failed to load class names from %s: %v", classesPath, err)
+		// Use default COCO class names as fallback
+		d.classes = cocoClassNames
+	}
+
+	// Try each model configuration
+	for _, model := range modelPaths {
+		if d.tryLoadModel(model.weights, model.config, model.name) {
+			d.initialized = true
+			return
+		}
+	}
+
+	log.Printf("Warning: No YOLO models found. OpenCV detection will not be available.")
+	d.initialized = false
+}
+
+// tryLoadModel attempts to load a specific YOLO model
+func (d *OpenCVDetector) tryLoadModel(weightsPath, configPath, modelName string) bool {
+	// Check if both files exist
+	if _, err := os.Stat(weightsPath); os.IsNotExist(err) {
+		return false
+	}
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
+		return false
+	}
+
+	// Try to load the network
+	net := gocv.ReadNetFromDarknet(configPath, weightsPath)
+	if net.Empty() {
+		log.Printf("Failed to load %s model from %s", modelName, weightsPath)
+		return false
+	}
+
+	// Set backend and target for optimal performance on Raspberry Pi
+	net.SetPreferableBackend(gocv.NetBackendOpenCV)
+	net.SetPreferableTarget(gocv.NetTargetCPU)
+
+	d.net = &net
+	log.Printf("Successfully loaded %s model for object detection", modelName)
+	return true
+}
+
+// loadClassNames loads COCO class names from file
+func (d *OpenCVDetector) loadClassNames(classesPath string) error {
+	data, err := os.ReadFile(classesPath)
+	if err != nil {
+		return err
+	}
+
+	classes := strings.Split(strings.TrimSpace(string(data)), "\n")
+	d.classes = classes
+	return nil
+}
+
+// DetectObjects performs object detection on an image file
+func (d *OpenCVDetector) DetectObjects(imagePath string) ([]DetectionBox, error) {
+	if !d.initialized || d.net == nil {
+		return nil, fmt.Errorf("OpenCV detector not properly initialized")
+	}
+
+	// Load the image in color mode (BGR) - YOLO models require 3-channel color input
+	// IMReadColor ensures we get BGR color data rather than grayscale, which is essential for accurate object detection
+	img := gocv.IMRead(imagePath, gocv.IMReadColor)
+	if img.Empty() {
+		return nil, fmt.Errorf("failed to load image: %s", imagePath)
+	}
+	defer img.Close()
+
+	return d.detectObjectsFromMat(img)
```

Comment:
image files are expected to contain 3-5MP photos. for detection & bounding box analysis, estimate the performance impact of taking a lower res photo such as 640x480 which is solely used for detection.

would this help OR is the in-memory downscaling to OpenCV's resolution efficient enough such that taking additional lower res photos is not needed